### PR TITLE
Implemented logging token usage (solves #322)

### PR DESCRIPTION
**Pull request by UmerHA**
Token usage is now tracked and logged into the file `memory/logs/token_usage`.
Solves #322

How it works:
- every time `ai.next` is called, the token usage is computed and tracked
- after the whole run, the token usage is formatted and printed to the file `memory/logs/token_usage`.

Output:
File in csv format with column:
- step_name
- prompt_tokens_in_step
- completion_tokens_in_step
- total_tokens_in_step
- total_prompt_tokens
- total_completion_tokens
- total_tokens

Notable changes:
- each step now passes its name to `ai.next`, which is used for logging
- added dependency to package `tiktoken` as it's needed to compute token usage when using streaming

screenshot-token-usage

### CODE REVIEW

**Comment by AntonOsika** [comment]
Thanks for this!

Two questions:
- Shouldn't we use the [response from the OpenAI API](https://platform.openai.com/docs/api-reference/chat/create) to compute the token usage? See screenshot.
- How can we solve the "name" of the steps with python? Can't merge it with hardcoded string for step name like this :/.  I think you should be able to ask chatgpt4. 

image

**Comment by UmerHA** [comment]
Hey @AntonOsika!

**Re 1:** Those usage stats are only available when you don't stream. When streaming, you have to calculate the token usage afterwards using tiktoken (see [this post on the openai forums](https://community.openai.com/t/openai-api-get-usage-tokens-in-response-when-set-stream-true/141866))

image


**Re 2:** Have changed it so the step names are now inferred from the function name via `step.__name__`. Does that work for you?

File: gpt_engineer/ai.py [hunk 1]
```diff
@@ -50,8 +65,64 @@ def next(self, messages: list[dict[str, str]], prompt=None):
...
+        prompt_tokens = self.num_tokens_from_messages(messages)
+        completion_tokens = self.num_tokens(answer)
+        total_tokens = prompt_tokens + completion_tokens
+
+        self.cumulative_prompt_tokens += prompt_tokens
+        self.cumulative_completion_tokens += completion_tokens
+        self.cumulative_total_tokens += total_tokens
+
+        self.token_usage_log.append({
+            "step_name": step_name,
```
Comment (AntonOsika) [inline gpt_engineer/ai.py:84]: Would be great to use a `dataclass` here to make it easier to get IDE support etc!

File: gpt_engineer/main.py [hunk 2]
```diff
@@ -61,11 +61,12 @@ def main(
 
     steps = STEPS[steps_config]
     for step in steps:
-        messages = step(ai, dbs)
+        messages = step(ai, dbs, step_name=step.__name__)
```
Comment (AntonOsika) [inline gpt_engineer/main.py:64]:
```suggestion
        messages = step(ai, dbs)
```

File: gpt_engineer/steps.py [hunk 3]
```diff
@@ -38,21 +38,21 @@ def get_prompt(dbs: DBs) -> str:
 # All steps below have the signature Step
 
 
-def simple_gen(ai: AI, dbs: DBs) -> List[dict]:
+def simple_gen(ai: AI, dbs: DBs, step_name:str) -> List[dict]:
```
Comment (AntonOsika) [inline gpt_engineer/steps.py:41]:
```suggestion
import inspect

def curr_fn() -> str:
    return inspect.currentframe().f_code.co_name

def simple_gen(ai: AI, dbs: DBs) -> List[dict]:
```

File: gpt_engineer/steps.py [hunk 4]
```diff
@@ -38,21 +38,21 @@ def get_prompt(dbs: DBs) -> str:
...
     """Run the AI on the main prompt and save the results"""
-    messages = ai.start(setup_sys_prompt(dbs), get_prompt(dbs))
+    messages = ai.start(setup_sys_prompt(dbs), get_prompt(dbs),step_name=step_name)
```
Comment (AntonOsika) [inline gpt_engineer/steps.py:43]:
```suggestion
    messages = ai.start(setup_sys_prompt(dbs), get_prompt(dbs), step_name=curr_fn())
```

**Review by AntonOsika — CHANGES_REQUESTED** [review]

**Comment by AntonOsika** [comment]
Makes sense!

Great job @UmerHA. Was about to say "let's merge" but I realized it can be done in a way that preserves what people complement this project for: How simple and clean it is.

If you would check my comments, resolve any conflicts, and run black (black should run if you setup pre-commit hooks like in [CONTRIBUTING.md](https://github.com/AntonOsika/gpt-engineer/blob/main/.github/CONTRIBUTING.md) I'd love to merge!

File: gpt_engineer/ai.py (hunk 1 above)
Comment (UmerHA) [reply gpt_engineer/ai.py]: agree - done ✅

File: gpt_engineer/steps.py (hunk 3 above)
Comment (UmerHA) [reply gpt_engineer/steps.py]: We need to use
```
def curr_fn() -> str:
    return inspect.stack()[1].function
```
to get the current function's name. Your suggestion would always return "curr_fn". But done ✅

**Comment by UmerHA** [comment]
@AntonOsika all done! Lmk if there's anything else!

File: gpt_engineer/ai.py [hunk 5]
```diff
@@ -31,7 +60,7 @@ def fuser(self, msg):
     def fassistant(self, msg):
         return {"role": "assistant", "content": msg}
 
-    def next(self, messages: List[Dict[str, str]], prompt=None):
+    def next(self, messages: list[Dict[str, str]], prompt=None, *, step_name=None):
```
Comment (AntonOsika) [inline gpt_engineer/ai.py:63]: Python 3.8 does not support the change to Lower case list here..:/

Comment (AntonOsika) [reply gpt_engineer/ai.py:63]:
```suggestion
    def next(self, messages: List[Dict[str, str]], prompt=None, *, step_name=None):
```

File: gpt_engineer/ai.py [hunk 6]
```diff
@@ -52,8 +81,65 @@ def next(self, messages: List[Dict[str, str]], prompt=None):
...
+                total_completion_tokens=self.cumulative_completion_tokens,
+                total_tokens=self.cumulative_total_tokens,
+            )
+        )
+
+    def format_token_usage_log(self):
+        result = "step_name,"
+        result += "prompt_tokens_in_step,completion_tokens_in_step,total_tokens_in_step"
+        result += ",total_prompt_tokens,total_completion_tokens,total_tokens\n"
+        for L in self.token_usage_log:
```
Comment (AntonOsika) [inline gpt_engineer/ai.py:116]: Prefer the lower case variable names a lot 😬

File: pyproject.toml [hunk 7]
```diff
@@ -19,6 +19,7 @@ dependencies = [
   'typer >= 0.3.2',
   'rudder-sdk-python == 2.0.2',
   'dataclasses-json == 0.5.7',
+  'tiktoken'
```
Comment (AntonOsika) [inline pyproject.toml:22]: Ah there should be a trailing comma here right?

File: gpt_engineer/ai.py (hunk 6 above)
Comment (UmerHA) [reply gpt_engineer/ai.py:116]: Haha, but ruff doesn't :D
I'll change the name to sth else

Comment (UmerHA) [reply gpt_engineer/ai.py]: change to `log`

**Comment by AntonOsika** [comment]
Fantastic

**Comment by AntonOsika** [comment]
Merging — just checking, did you test this runs as expected, and if not could you do it? (Traveling for work atm)

**Comment by UmerHA** [comment]
> Merging — just checking, did you test this runs as expected, and if not could you do it? (Traveling for work atm)

Of course I did :)

**Pull request merged**